README.rst
***********
sc.embedder
***********

.. contents::

Life, the Universe, and Everything
----------------------------------

This packages contains a Dexterity-based content type that allows you to
embedded content (such as photos or videos) from third parties into your Plone
site.

You can easily embed content from Flickr, SlideShare, SoundCloud, Vimeo,
Wikipedia, YouTube, and many other.

Mostly Harmless
---------------

.. image:: https://secure.travis-ci.org/simplesconsultoria/sc.embedder.png?branch=master
    :target: http://travis-ci.org/simplesconsultoria/sc.embedder

.. image:: https://coveralls.io/repos/simplesconsultoria/sc.embedder/badge.png
    :target: https://coveralls.io/r/simplesconsultoria/sc.embedder?branch=master

Got an idea? Found a bug? Let us know by `opening a support ticket`_.

Don't Panic
-----------

Installation
^^^^^^^^^^^^

To enable this product in a buildout-based installation:

1. Edit your buildout.cfg and add ``sc.embedder`` to the list of eggs to
   install ::

    [buildout]
    ...
    eggs =
        sc.embedder

2. If you are using Plone 4.1 you may need to extend a Dexterity known good
   set (KGS) to make sure that you get the right versions of the packages that
   make up Dexterity::

    [buildout]
    ...
    extends =
        http://good-py.appspot.com/release/dexterity/1.2.1

After updating the configuration you need to run ''bin/buildout'', which will
take care of updating your system.

Go to the 'Site Setup' page in a Plone site and click on the 'Add-ons' link.

Check the box next to `sc.embedder`` and click the 'Activate' button.

.. Note::

	You may have to empty your browser cache and save your resource registries
	in order to see the effects of the product installation.

Use
^^^

Once the product is installed you will note that you are available to add a
Embedder content type. When you attempt to add an item you will see the
fields:

- Content URL
- Description
- Embed html code
- Width
- Height
- Player position
- Body text
- Alternate content
- Image

In the URL field you must put the url of you content. Like:

    http://vimeo.com/17914974

and then click the *load* button. This will bring automatically some data of
the element from the provider to the form. It depends on each provider what
data it going to add. But generally the fields title, description, width,
height and embed html fields are the one to be filled.

**Note**: *Some providers only support width, height and title. So the fields
like description or html embed code are necessarily to be filled manually.
This information is normally encountered in the page where the url is pointing
to*.

Layout
^^^^^^
With *player position* field you can set where the embed multimedia is going
to be render regarding the body text. You have the options *top, bottom, left
and right*.

Body text will be the main information that will describe the content.

The *alternate content* is to give a description or transcription to an
individual that is no able to hear. In the layout this is hidden by default.
There is a link called **alternate content** that will show or hide again the
content of this field by clicking on it.

The *Image* field is just to assign an image to the object that will give a
quick visual description of the content. The idea is to show it in listing
views or covers.

.. _`opening a support ticket`: https://github.com/simplesconsultoria/sc.embedder/issues


CHANGES.rst
Changelog
---------

There's a frood who really knows where his towel is.

1.5.1 (unreleased)
^^^^^^^^^^^^^^^^^^

- Fix error 403 for Youtube oembed endpoint (Ref. `#95`_).
  [idgserpro]

1.5.0 (2019-05-07)
^^^^^^^^^^^^^^^^^^

- Allows Embedder content to be linkcable in TinyMCE.
  [idgserpro]

- Add ``cssselect`` as pakage dependency.
  [idgserpro]


1.5b2 (2017-10-10)
^^^^^^^^^^^^^^^^^^

- Code clean up;
  enforce usage of Products.PortalTransforms >=2.1.10 as ``<iframe>`` must be listed as a valid tag.
  [hvelarde]

- Remove hard dependency on collective.dexteritytextindexer.
  [hvelarde]

- Update ``SearchableText`` index for Embedder objects.
  [hvelarde]


1.5b1 (2017-09-20)
^^^^^^^^^^^^^^^^^^

- Fix i18n and update Brazilian Portuguese and Spanish translations.
  [rafahela, hvelarde]

- ``SearchableText`` index for Embedder content type now includes object id and keywords.
  [hvelarde]

- Remove dependency on collective.dexteritytextindexer.
  [hvelarde]


1.4b1 (2017-09-06)
^^^^^^^^^^^^^^^^^^

- Add Related Items behavior to Embedder content type by default (closes `#80 <https://github.com/simplesconsultoria/sc.embedder/issues/80>`_).
  [hvelarde]

- Add Embedder to TinyMCE's list of linkable content types (closes `#79 <https://github.com/simplesconsultoria/sc.embedder/issues/79>`_).
  [hvelarde]

- Remove hard dependency on Products.TinyMCE.
  [hvelarde]

- Drop support for Plone 4.2.
  [hvelarde]


1.3b1 (2017-02-20)
^^^^^^^^^^^^^^^^^^

- Use semantic markup on the Embedder view (for simplicity, we are using `WebPage <http://schema.org/WebPage>`_  type).
  [hvelarde]

- Remove references to Grok's ``w`` dictionary used on the view template (fixes `#72`_).
  [rodfersou]


1.2b2 (2017-01-09)
^^^^^^^^^^^^^^^^^^

- Update list of package dependencies.
  [hvelarde]

- Fix width validator.
  [hvelarde]

- Remove dependency on plone.directives.form and latest traces of Grok.
  [hvelarde]


1.2b1 (2016-12-16)
^^^^^^^^^^^^^^^^^^

- Remove dependency on five.grok (closes `#63`_).
  [rodfersou]

- Package is compatible with Plone 4.2 again.
  [hvelarde]

- Avoid `TypeError` while running upgrade step (fixes `#58`_).
  [hvelarde]


1.1b2 (2016-04-25)
^^^^^^^^^^^^^^^^^^

- Implement `Subresource Integrity <https://www.w3.org/TR/SRI/>`_ (closes `#55`_).
  [cleberjsantos]

- Sanitize <iframe> tags to avoid including invalid HTML attributes;
  an upgrade step is included to clean up existing objects (closes `#44`_).
  [rodfersou, hvelarde]

- Update Video.js to v5.8 and load it from CDN (closes `#50`_).
  [rodfersou]


1.1b1 (2016-03-24)
^^^^^^^^^^^^^^^^^^

- Remove hard dependency on plone.app.referenceablebehavior as Archetypes is no longer the default framework in Plone 5.
  Under Plone < 5.0 you should now explicitly add it to the `eggs` part of your buildout configuration to avoid issues while upgrading.
  [hvelarde]

- Fixes issue in TinyMCE plugin when clicked in Insert button (closes `#16`_).
  [idgserpro]

- Fix error in TinyMCE plugin if the property icon_visibility is disabled (closes `#42`_).
  [rodfersou]


1.0b5 (2015-11-18)
^^^^^^^^^^^^^^^^^^

- Fix the stripping of embed code when it has more than one tag, like in Facebook videos (closes `#39`_).
  [rodfersou]

- Use "application/javascript" media type instead of the obsolete "text/javascript".
  [hvelarde]

- Remove Chrome Frame from ``X-UA-Compatible`` HTTP header as it's deprecated.
  [hvelarde]

- Allow use of percent sign (%) on width properties (closes `#6`_).
  [rodfersou, hvelarde]

- Added italian translation
  [keul]


1.0b4 (2015-09-03)
^^^^^^^^^^^^^^^^^^

- Add Embedder tile for collective.cover (closes `#32`_).
  [rodfersou]

- Brazilian Portuguese and Spanish translations were updated.
  [rodfersou, hvelarde]

- Add portal message instead of log info on HTTP request errors (closes `#14`_).
  [rodfersou]

- Add validation in URL field to avoid hiding errors when using the Load button (closes `#20`_).
  [rodfersou]

- Remove dependency on collective.oembed (closes `#3`_).
  [hvelarde]

- Drop support for Plone 4.1 and 4.2; remove dependency on unittest2.
  [hvelarde]


1.0b3 (2013-07-23)
^^^^^^^^^^^^^^^^^^

- Fix a couple AJAX quoting/unquoting problems on the TinyMCE plugin.
  [jsbueno]

- Add helper methods image_thumb and tag in order to be listed in
  folder_summary_view [ericof]

- Fix an UnicodeDecodeError with our plugin for TinyMCE [ericof]


1.0b2 (2012-12-03)
^^^^^^^^^^^^^^^^^^

- Fix a conflict with plone.formwidget.namedfile NamedImage widget
  implementation. [jpgimenez]


1.0b1 (2012-11-27)
^^^^^^^^^^^^^^^^^^

- Update package dependecies for Plone 4.3 compatibility. [hvelarde]

- Fix functional tests. [hvelarde]

- Rename package: was sc.content.embedder and now is sc.embedder. [hvelarde]

- Support for VideoJS as fallback if static file and no supported provider.
  [jpgimenez]

- Fixed the rendering of the embedded code to not break the main view.
  [jpgimenez]

- Allow selecting, embedding and rendering sc.embedder content into TinyMCE
  widgets as if it were images. [jpgimenez]

- VideoJS embedder code implemented as an iframe, to simplify the integration
  with TinyMCE. [jpgimenez]


1.0a3 (2012-10-04)
^^^^^^^^^^^^^^^^^^

- Fixed KeyError: 'width' when saving embeded HTML with percentages.
  [davilima6]


1.0a2 (2012-09-05)
^^^^^^^^^^^^^^^^^^

- Functional tests were updated to run with robotframework-selenium2library.
  [hvelarde]

- i18n was fixed and Spanish translation was updated. [hvelarde]

- Brazilian Portuguese translation was fixed. [agnogueira]


1.0a1 (2012-09-05)
^^^^^^^^^^^^^^^^^^^

- Initial release.

.. _`#3`: https://github.com/simplesconsultoria/sc.embedder/issues/3
.. _`#6`: https://github.com/simplesconsultoria/sc.embedder/issues/6
.. _`#14`: https://github.com/simplesconsultoria/sc.embedder/issues/14
.. _`#16`: https://github.com/simplesconsultoria/sc.embedder/issues/16
.. _`#20`: https://github.com/simplesconsultoria/sc.embedder/issues/20

src/sc/embedder/tests/test_setup.py
# -*- coding: utf-8 -*-
from plone import api
from plone.browserlayer.utils import registered_layers
from sc.embedder.config import PROJECTNAME
from sc.embedder.testing import INTEGRATION_TESTING
from sc.embedder.testing import IS_PLONE_5

import unittest


class InstallTestCase(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.qi = self.portal['portal_quickinstaller']

    def test_installed(self):
        self.assertTrue(self.qi.isProductInstalled(PROJECTNAME))

    def test_browserlayer(self):
        layers = [l.getName() for l in registered_layers()]
        self.assertIn('IEmbedderLayer', layers)

    @unittest.skipIf(IS_PLONE_5, 'No easy way to test this under Plone 5')
    def test_css_registry(self):
        resource_ids = self.portal.portal_css.getResourceIds()
        self.assertIn('embedder.css', resource_ids)

    def test_add_permissions(self):
        permission = 'sc.embedder: Add Embedder'
        roles = self.portal.rolesOfPermission(permission)
        roles = [r['name'] for r in roles if r['selected']]
        expected = ['Contributor', 'Manager', 'Owner', 'Site Administrator']
        self.assertListEqual(roles, expected)

    def test_tinymce_linkables(self):
        linkables = self.portal.portal_tinymce.linkable.split('\n')
        self.assertIn('sc.embedder', linkables)

    def test_tile(self):
        tiles = api.portal.get_registry_record('plone.app.tiles')
        self.assertIn(u'sc.embedder', tiles)


class UninstallTest(unittest.TestCase):

    layer = INTEGRATION_TESTING

    def setUp(self):
        self.portal = self.layer['portal']
        self.qi = self.portal['portal_quickinstaller']
        self.qi.uninstallProducts(products=[PROJECTNAME])

    def test_uninstalled(self):
        self.assertFalse(self.qi.isProductInstalled(PROJECTNAME))

    def test_browserlayer_removed(self):
        layers = [l.getName() for l in registered_layers()]
        self.assertNotIn('IEmbedderLayer', layers)

    @unittest.skipIf(IS_PLONE_5, 'No easy way to test this under Plone 5')
    def test_css_removed(self):
        resource_ids = self.portal.portal_css.getResourceIds()
        self.assertNotIn('embedder.css', resource_ids)

    def test_tinymce_linkables(self):
        linkables = self.portal.portal_tinymce.linkable.split('\n')
        self.assertNotIn('sc.embedder', linkables)

    def test_tile_removed(self):
        tiles = api.portal.get_registry_record('plone.app.tiles')
        self.assertNotIn(u'sc.embedder', tiles)


src/sc/embedder/tests/test_adapters.py
# -*- coding: utf-8 -*-
"""Module with the apapters tests."""

from plone import api
from sc.embedder.testing import INTEGRATION_TESTING
from sc.embedder.testing import IS_PLONE_5

import unittest


class AdaptersTestCase(unittest.TestCase):
    """Adapters test class."""

    layer = INTEGRATION_TESTING

    def setUp(self):
        """Tests configuration."""
        self.portal = self.layer['portal']
        with api.env.adopt_roles(['Manager']):
            self.multimedia = api.content.create(
                self.portal,
                'sc.embedder',
                title='Multimedia',
                description='New Multimedia',
                embed_html='<iframe width="480" height="270" '
                'src="https://www.teste.com/embed/tesste?feature=oembed" '
                'allowfullscreen></iframe>',
                width=300,
                height=200,
            )

    @unittest.skipIf(IS_PLONE_5, 'Plone 5 has no Products.TinyMCE.')
    def test_json_details(self):
        """Test if JSONDetails adapter returns all keys."""
        from Products.TinyMCE.adapters.interfaces.JSONDetails import IJSONDetails
        adapter = IJSONDetails(self.multimedia, None)
        uid = self.multimedia.UID()
        expected = '{{"thumb_html": "%3Ciframe%20width%3D%22188%22%20height'\
            '%3D%22141%22%20src%3D%22https%3A//www.teste.com/embed/tesste%3F'\
            'feature%3Doembed%22%20allowfullscreen%3E%3C/iframe%3E",'\
            ' "uid_relative_url": "resolveuid/{0}", "thumb": "", "uid_url": '\
            '"http://nohost/plone/resolveuid/{0}", "url": '\
            '"http://nohost/plone/multimedia", "description": '\
            '"New Multimedia", "title": "Multimedia", "embed_html": '\
            '"%3Ciframe%20width%3D%22480%22%20height'\
            '%3D%22270%22%20src%3D%22https%3A//www.teste.com/embed/'\
            'tesste%3Ffeature%3Doembed%22%20allowfullscreen%3E%3C/'\
            'iframe%3E", "anchors": []}}'.format(uid)
        result = adapter.getDetails()
        self.assertEqual(expected, result)


src/sc/embedder/adapters.py
# -*- coding: utf-8 -*-
from lxml import cssselect
from lxml import etree
from lxml import html
from Products.TinyMCE.adapters.interfaces.JSONDetails import IJSONDetails
from Products.TinyMCE.adapters.JSONDetails import JSONDetails
from urllib2 import quote
from zope.interface import implementer

import json


@implementer(IJSONDetails)
class JSONDetails(JSONDetails):
    """Return details of the current object in JSON"""

    def getDetails(self):
        """Builds a JSON object based on the details
           of this object.
        """

        results = json.loads(super(JSONDetails, self).getDetails())

        results['description'] = self.context.description

        tree = etree.HTML(self.context.embed_html)
        sel = cssselect.CSSSelector('body > *')
        el = sel(tree)[0]
        el.attrib['width'] = '188'
        el.attrib['height'] = '141'

        # The raw html bits don't travel nicely over json alone
        # use javascript's decodeURIComponent on the client side
        # to get these values back:
        results['thumb_html'] = quote(html.tostring(el))
        results['embed_html'] = quote(self.context.embed_html)

        return json.dumps(results)


src/sc/embedder/content/embedder.py
# -*- coding: utf-8 -*-
from lxml import cssselect
from lxml import etree
from lxml import html
from lxml.html.builder import DIV
from plone import api
from plone.app.textfield import RichText
from plone.app.textfield.interfaces import ITransformer
from plone.autoform import directives as form
from plone.dexterity.browser.add import DefaultAddForm
from plone.dexterity.browser.add import DefaultAddView
from plone.dexterity.browser.base import DexterityExtensibleForm
from plone.dexterity.browser.edit import DefaultEditForm
from plone.dexterity.content import Item
from plone.dexterity.events import AddCancelledEvent
from plone.dexterity.events import EditCancelledEvent
from plone.dexterity.events import EditFinishedEvent
from plone.formwidget.namedfile.widget import NamedImageWidget
from plone.indexer import indexer
from plone.namedfile.field import NamedImage as BaseNamedImage
from plone.namedfile.file import NamedImage as ImageValueType
from plone.supermodel import model
from Products.CMFPlone.utils import safe_unicode
from Products.Five.browser import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile
from sc.embedder import MessageFactory as _
from sc.embedder.interfaces import IConsumer
from sc.embedder.logger import logger
from sc.embedder.utils import sanitize_iframe_tag
from sc.embedder.utils import validate_int_or_percentage
from urllib2 import HTTPError
from urllib2 import URLError
from z3c.form import button
from z3c.form.interfaces import IFieldWidget
from z3c.form.interfaces import IFormLayer
from z3c.form.widget import FieldWidget
from zope import component
from zope import schema
from zope.component import adapter
from zope.event import notify
from zope.interface import implementer
from zope.schema.vocabulary import SimpleVocabulary
from zope.schema.vocabulary import SimpleTerm

import requests
import urllib2
import urlparse


def has_image(content):
    image = content.image
    return (image and image.getSize())


class EmbedderImageWidget(NamedImageWidget):

    klass = u'embedder-image-widget'

    @property
    def download_url(self):
        if self.field is None:
            return None
        if getattr(self, 'url', None):
            return self.url
        if self.ignoreContext:
            return None
        if self.filename_encoded:
            return '%s/++widget++%s/@@download/%s' % (self.request.getURL(),
                                                      self.name,
                                                      self.filename_encoded)
        else:
            return '%s/++widget++%s/@@download' % (self.request.getURL(),
                                                   self.name)


class NamedImage(BaseNamedImage):
    """
    Do not conflict with plone.formwidget.namedfile widget implementation.
    """


@implementer(IFieldWidget)
@adapter(NamedImage, IFormLayer)
def EmbedderImageFieldWidget(field, request):
    return FieldWidget(field, EmbedderImageWidget(request))


class IEmbedder(model.Schema):
    """ A representation of a content embedder content type
    """

    form.order_before(**{'url': '*'})

    url = schema.ASCIILine(
        title=_(u'Multimedia URL'),
        description=_(u'The URL for your multimedia file. Can be a URL ' +
                      u'from YouTube, Vimeo, SlideShare, SoundCloud or ' +
                      u'other main multimedia websites.'),
        required=False,
    )

    width = schema.ASCIILine(
        title=_(u'Width'),
        description=_(
            u'Can be expressed as a decimal number or a percentage, e.g., 270 or 50%.'),
        required=True,
        constraint=validate_int_or_percentage,
    )

    height = schema.Int(
        title=_(u'Height'),
        description=_(u'Can be expressed as a decimal number, e.g., 480.'),
        required=True,
        min=1,
        max=9999,
    )

    embed_html = schema.Text(
        title=_(u'Embedded HTML code'),
        description=_(u'HTML code to render the embedded media.'),
        required=True,
    )

    player_position = schema.Choice(
        title=_(u'Player position'),
        description=_(u''),
        default=u'Top',
        required=True,
        vocabulary=SimpleVocabulary([
            SimpleTerm(value=u'Top', title=_(u'Top')),
            SimpleTerm(value=u'Bottom', title=_(u'Bottom')),
            SimpleTerm(value=u'Left', title=_(u'Left')),
            SimpleTerm(value=u'Right', title=_(u'Right'))
        ])
    )

    form.widget(image=EmbedderImageFieldWidget)
    image = NamedImage(
        title=_(u'Preview image'),
        description=_(u'Image to be used when listing content.'),
        required=False,
    )

    text = RichText(
        title=_(u'Body text'),
        required=False,
    )

    alternate_content = RichText(
        title=_(u'Alternate content'),
        description=_(u'Description or transcription for accessibility (WCAG) users.'),
        required=False,
    )


@implementer(IEmbedder)
class Embedder(Item):

    """A content embedder."""

    def image_thumb(self):
        """Return a thumbnail."""

        if not has_image(self):
            return None
        view = self.unrestrictedTraverse('@@images')
        return view.scale(fieldname='image',
                          scale='thumb').index_html()

    def tag(self, scale='thumb', css_class='tileImage', **kw):
        """Return a tag to the image."""

        if not (has_image(self)):
            return ''
        view = self.unrestrictedTraverse('@@images')
        return view.tag(fieldname='image',
                        scale=scale,
                        css_class=css_class,
                        **kw)


class BaseForm(DexterityExtensibleForm):

    """Methods and attributes shared by Add and Edit form."""

    tr_fields = {'width': 'width',
                 'height': 'height',
                 'description': 'IDublinCore.description',
                 'title': 'IDublinCore.title',
                 'html': 'embed_html'}

    def set_image(self, url):
        opener = urllib2.build_opener()
        try:
            response = opener.open(url)
            self.widgets['image'].url = url
            self.widgets['image'].value = ImageValueType(data=response.read(),
                                                         filename=url.split('/')[-1])
            self.request['form.widgets.image.action'] = u'load'
        except:
            pass

    def get_data(self, url, maxwidth=None, maxheight=None, format='json'):
        """Return the data provided by the endpoint."""

        consumer = component.getUtility(IConsumer)
        json_data = None
        try:
            json_data = consumer.get_data(
                url, maxwidth=None, maxheight=None, format='json')
        except HTTPError, e:
            if e.code == 401:
                api.portal.show_message(
                    _(u'Unauthorized request'), request=self.request, type='error')
            elif e.code == 404:
                api.portal.show_message(
                    _(u'URL not found'), request=self.request, type='error')
            else:
                logger.warn(e)
        except URLError, e:
            # support offline mode
            logger.warn('offline mode')

        return json_data

    def handle_image(self, data):
        url = self.widgets['url'].value
        action = self.request.get('form.widgets.image.action', None)
        if action == 'load':
            json_data = self.get_data(
                url, maxwidth=None, maxheight=None, format='json')
            if json_data.get('thumbnail_url'):
                opener = urllib2.build_opener()
                try:
                    response = opener.open(json_data.get('thumbnail_url'))
                    data['image'] = ImageValueType(data=response.read(),
                                                   filename=json_data.get('thumbnail_url').split('/')[-1])
                except:
                    pass

    def get_fallback(self, url):
        supported_mime_types = ('video/mp4', 'video/ogg', 'video/webm')
        embedder_code = """
<iframe src="%(context_url)s/@@embedder_videojs?src=%(url)s&type=%(type)s"
        class="vjs-iframe"
        allowfullscreen>
</iframe>
"""
        r = requests.head(url)
        if r.headers.get('content-type') in supported_mime_types:
            return {'html': embedder_code % {'context_url': self.context.absolute_url(),
                                             'url': urllib2.quote(url, ''),
                                             'type': urllib2.quote(r.headers.get('content-type'), '')}}

    def _validate_url(self, url):
        """Validate if is a valid URL for site.

        :param url: [required] URL to validate.
        :type url: unicode
        :returns: True if valid site URL
        :rtype: bool
        """
        pieces = urlparse.urlparse(url)
        # Must have a net location (ex: www.youtube.com)
        if pieces.netloc == '':
            return False
        if pieces.scheme not in ('http', 'https'):
            return False
        return True

    def load_oembed(self, action):
        url = self.widgets['url'].value

        if not self._validate_url(url):
            api.portal.show_message(
                _(u'Invalid URL'), request=self.request, type='error')
            return

        json_data = self.get_data(
            url, maxwidth=None, maxheight=None, format='json')

        if json_data is None:
            json_data = self.get_fallback(url)
            if json_data is None:
                return
        # html parameter not always required:
        # https://github.com/abarmat/python-oembed/blob/master/oembed/__init__.py#L157-L167
        # https://github.com/abarmat/python-oembed/blob/master/oembed/__init__.py#L181-L187
        if 'html' in json_data:
            json_data['html'] = sanitize_iframe_tag(json_data['html'])
        for k, v in self.tr_fields.iteritems():
            if json_data.get(k):
                self.widgets[v].value = unicode(json_data[k])
        if json_data.get('thumbnail_url'):
            self.set_image(json_data.get('thumbnail_url'))

    def set_custom_embed_code(self, data):
        """Return the code that embed the code.

        Could be with the original size or the custom chosen.
        """
        if 'embed_html' not in data:
            return
        tree = etree.HTML(data['embed_html'])
        sel = cssselect.CSSSelector('body > *')
        el = sel(tree)
        # add a div around if there is more than one element into code
        if len(el) > 1:
            el = DIV(*el)
        else:
            el = el[0]

        # width and height attributes should not be set in a div tag
        if el.tag in ['iframe', 'object']:
            if data.get('width', None):
                el.attrib['width'] = data['width'] and str(data['width']) or el.attrib['width']
            if data.get('height', None):
                el.attrib['height'] = data['height'] and str(data['height']) or el.attrib['height']

        data['embed_html'] = sanitize_iframe_tag(html.tostring(el))


class AddForm(BaseForm, DefaultAddForm):
    template = ViewPageTemplateFile('templates/sc.embedder.pt')

    @button.buttonAndHandler(_('Save'), name='save')
    def handleAdd(self, action):
        if self.request.get('form.widgets.url') and \
           not self.request.get('form.widgets.embed_html'):
            fields = {'width': 'form.widgets.width',
                      'height': 'form.widgets.height',
                      'description': 'form.widgets.IDublinCore.description',
                      'title': 'form.widgets.IDublinCore.title',
                      'html': 'form.widgets.embed_html'}
            consumer = component.getUtility(IConsumer)
            json_data = consumer.get_data(self.request['form.widgets.url'],
                                          maxwidth=None,
                                          maxheight=None,
                                          format='json')
            for k, v in fields.iteritems():
                if json_data.get(k):
                    self.request[v] = unicode(json_data[k])
        data, errors = self.extractData()
        self.handle_image(data)
        self.set_custom_embed_code(data)
        if errors:
            self.status = self.formErrorsMessage
            return
        obj = self.createAndAdd(data)
        if obj is not None:
            # mark only as finished if we get the new object
            self._finishedAdd = True
            api.portal.show_message(
                _(u'Item created'), self.request, type='info')

    @button.buttonAndHandler(_(u'Cancel'), name='cancel')
    def handleCancel(self, action):
        api.portal.show_message(
            _(u'Add New Item operation cancelled.'), self.request, type='info')
        self.request.response.redirect(self.nextURL())
        notify(AddCancelledEvent(self.context))

    @button.buttonAndHandler(_('Load'), name='load')
    def handleLoad(self, action):
        self.load_oembed(action)

    def get_url_widget(self):
        widget = [key for key in self.widgets.values()
                  if key.id == 'form-widgets-url']
        if widget != []:
            url_w = widget[0]
            return url_w

    def get_load_action(self):
        action = [key for key in self.actions.values()
                  if key.id == 'form-buttons-load']
        if action != []:
            load = action[0]
            return load


class AddView(DefaultAddView):
    form = AddForm


class EditForm(DefaultEditForm, BaseForm):
    template = ViewPageTemplateFile('templates/edit.pt')

    @button.buttonAndHandler(_('Load'), name='load')
    def handleLoad(self, action):
        self.load_oembed(action)

    @button.buttonAndHandler(_(u'Save'), name='save')
    def handleApply(self, action):
        data, errors = self.extractData()
        self.handle_image(data)
        self.set_custom_embed_code(data)
        if errors:
            self.status = self.formErrorsMessage
            return
        self.applyChanges(data)
        api.portal.show_message(
            _(u'Changes saved.'), self.request, type='info')
        self.request.response.redirect(self.nextURL())
        notify(EditFinishedEvent(self.context))

    @button.buttonAndHandler(_(u'Cancel'), name='cancel')
    def handleCancel(self, action):
        api.portal.show_message(
            _(u'Edit cancelled.'), self.request, type='info')
        self.request.response.redirect(self.nextURL())
        notify(EditCancelledEvent(self.context))

    def get_url_widget(self):
        widget = [key for key in self.widgets.values()
                  if key.id == 'form-widgets-url']
        if widget != []:
            url_w = widget[0]
            return url_w

    def get_load_action(self):
        action = [key for key in self.actions.values()
                  if key.id == 'form-buttons-load']
        if action != []:
            load = action[0]
            return load


class View(BrowserView):
    index = ViewPageTemplateFile('templates/view.pt')

    def get_player_pos_class(self):
        """ Returns the css class based on the position of the embed item.
        """
        pos = self.context.player_position
        css_class = '%s_embedded' % pos.lower()
        return css_class

    def __call__(self):
        return self.index()


class EmbedderVideoJS(BrowserView):
    index = ViewPageTemplateFile('templates/embeddervideojs.pt')

    def __call__(self):
        return self.index()


@indexer(IEmbedder)
def searchable_text_indexer(obj):
    """SearchableText should contain id, title, description, body text,
    alternate text and keywords.
    """
    transformer = ITransformer(obj)

    try:
        text = transformer(obj.text, 'text/plain')
    except AttributeError:
        text = ''

    try:
        alternate_text = transformer(obj.alternate_content, 'text/plain')
    except AttributeError:
        alternate_text = ''

    keywords = u' '.join(safe_unicode(s) for s in obj.Subject())

    return u' '.join((
        safe_unicode(obj.id),
        safe_unicode(obj.title) or u'',
        safe_unicode(obj.description) or u'',
        safe_unicode(text),
        safe_unicode(alternate_text),
        safe_unicode(keywords),
    ))
